include/mlboard_logger.hpp

```
@@ -0,0 +1,200 @@
+  }
+
+  /**
+   * Callback function called at the end of a pass over the data.
+   *
+   * @param optimizer The optimizer used to update the function.
+   * @param function Function to optimize.
+   * @param coordinates Starting point.
+   * @param epoch The index of the current epoch.
+   * @param objective Objective value of the current point.
+   */
+  template<typename OptimizerType, typename FunctionType, typename MatType>
+  void EndEpoch(OptimizerType& /* optimizer */,
+                FunctionType& /* function */,
+                const MatType& /* coordinates */,
+                const size_t epoch,
+                double objective)
+  {
+    if (callbackUsed && summaryType == "")
+    {
+      objective = localFunc();
+    }
+    if (summaryType == "")
+    {
+      if (epoch % epochCount == 0)
+      {
+        mlboard::SummaryWriter<mlboard::FileWriter>::Scalar(lossTag,
+            epoch / epochCount, objective, output);
+        mlboard::SummaryWriter<mlboard::FileWriter>::Scalar(accTag,
+            epoch / epochCount, 1 - objective, output);
+      }
+    }
+    if (callbackUsed && (epoch % epochCount == 0) && summaryType != "")
+    {
+      arma::mat valueToBeLogged = matFunc();
+      if (summaryType == "histogram")
+      {
+        mlboard::SummaryWriter<mlboard::FileWriter>::Histogram(summaryTag,
+            epoch / epochCount, arma::rowvec(valueToBeLogged), output);
+      }
+      else if (summaryType == "embedding")
+      {
+        mlboard::SummaryWriter<mlboard::FileWriter>::Embedding(summaryTag,
+            valueToBeLogged, embeddingMetadata, output);
+      }
+      else if (summaryType == "image")
+      {
+        // channel is not needed to log image
+        mlpack::data::ImageInfo info(widthofImage, heightofImage, 0);
+        mlboard::SummaryWriter<mlboard::FileWriter>::Image(
+              summaryTag, epoch / epochCount , valueToBeLogged,
+              info, output);
+      }
+      else
+      {
+        /* Do nothing */
```

I think instead of leaving it like this, we can raise an error that this summary type is not supported by the callback.